### PR TITLE
Add dockerfile for cross itself

### PR DESCRIPTION
**Pull request by ByteDream**
This adds a dockerfile which has cross and docker already installed and can be directly used to cross compile and test. It mostly adopts a comment in [#260](https://github.com/cross-rs/cross/issues/260#issuecomment-678829526). 

It uses docker in docker (#387) so to work correctly it must be run with the docker socket mounted as volume.
```
$ docker run -v /var/run/docker.sock:/var/run/docker.sock cross
root@1e27ac2ed941:/# git clone https://github.com/cross-rs/cross
root@1e27ac2ed941:/# cd cross
root@1e27ac2ed941:/cross# cross build --target aarch64-unknown-linux-gnu
```

Although the [README already describes](https://github.com/cross-rs/cross#docker-in-docker) how to achieve this, a provided Dockerfile that can be downloaded from Docker Hub in the best case would be much easier than writing your own. Especially when using a CI pipeline inside Docker, where creating a custom Dockerfile is not that easy, this could be a big help.
The directory of the Dockerfile may not be the best choice since only compile targets are in this directory but it was the best fit.

### CODE REVIEW

File: docker/Dockerfile.cross [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM rust:latest
+
+ENV CROSS_DOCKER_IN_DOCKER=true
+
+RUN wget -O- https://download.docker.com/linux/static/stable/x86_64/docker-20.10.9.tgz | tar -C /usr/bin/ --strip-components=1 -xzf - docker/docker
```
Comment (otavio) [inline docker/Dockerfile.cross:5]: We need to check signature or checksum here so we don't rely on the file being unmodified in future.

**Review by otavio — CHANGES_REQUESTED** [review]

File: docker/Dockerfile.cross (hunk 1 above)
Comment (Emilgardis) [reply docker/Dockerfile.cross:5]: I don't think this is necessary.

Comment (otavio) [reply docker/Dockerfile.cross:5]: It is not rare for tarball files to change behind the scene, so ensuring it is valid is a good security matter.

**Comment by ByteDream** [comment]
The checksum check is now implemented (md5) and it only checks the `docker` binary inside of the tarball since this is the only file which is needed.

**Review by otavio — DISMISSED** [review]

File: docker/Dockerfile.cross [hunk 2]
```diff
@@ -0,0 +1,10 @@
+FROM rust:latest
+
+ENV CROSS_DOCKER_IN_DOCKER=true
+
+ARG checksum="d9d9b964d428733e98c2579b91772af4  /usr/bin/docker"
+
+RUN wget -O- https://download.docker.com/linux/static/stable/x86_64/docker-20.10.9.tgz | tar -C /usr/bin/ --strip-components=1 -xzf - docker/docker
+RUN echo $checksum | md5sum -c
+
+RUN cargo install cross
```
Comment (reitermarkus) [inline docker/Dockerfile.cross:10]: This should install the current version in this repository.

File: docker/Dockerfile.cross (hunk 1 above)
Comment (reitermarkus) [reply docker/Dockerfile.cross:7]: Is there a Docker image we can copy the binary from instead using a multi-stage build?

Comment (ByteDream) [reply docker/Dockerfile.cross:7]: There is a [docker in docker image](https://hub.docker.com/_/docker,) but it has some other binaries which are not necessary and rust has to be installed manually which takes more steps than this multi-stage build.
The docker in docker image receives it binaries from `download.docker.com` ([layer 6](https://hub.docker.com/layers/docker/library/docker/latest/images/sha256-3e1787157b0cd05a6c56f1d7bc916aa486f53f40f0d1538e34b917bdf947f94d)) just like here, so there is no big difference when using the docker in docker image (besides the unwanted binaries which it has).

File: docker/Dockerfile.cross [hunk 3]
```diff
@@ -7,4 +7,4 @@ ARG checksum="d9d9b964d428733e98c2579b91772af4  /usr/bin/docker"
 RUN wget -O- https://download.docker.com/linux/static/stable/x86_64/docker-20.10.9.tgz | tar -C /usr/bin/ --strip-components=1 -xzf - docker/docker
 RUN echo $checksum | md5sum -c
 
-RUN cargo install cross
+RUN cargo install --git https://github.com/cross-rs/cross
```
Comment (Emilgardis) [inline docker/Dockerfile.cross:10]: This is still wrong imo, I'd expect this to be built in our CI and published as `ghcr.io/cross-rs/cross-in-docker`.

**Review by Emilgardis — CHANGES_REQUESTED** [review]

File: docker/Dockerfile.cross (hunk 3 above)
Comment (reitermarkus) [reply docker/Dockerfile.cross]: Yes, this needs to be something like

```
COPY . .
RUN cargo install --path . --locked
```

**Comment by otavio** [comment]
@ByteDream could you rebase and add a changelog entry? So we could trigger CI and see if it'll succeed.

File: docker/Dockerfile.cross-in-docker [hunk 4]
```diff
@@ -0,0 +1,10 @@
+FROM rust:latest
+
+ENV CROSS_DOCKER_IN_DOCKER=true
+
+ARG checksum="d9d9b964d428733e98c2579b91772af4  /usr/bin/docker"
+
+RUN wget -O- https://download.docker.com/linux/static/stable/x86_64/docker-20.10.9.tgz | tar -C /usr/bin/ --strip-components=1 -xzf - docker/docker
+RUN echo $checksum | md5sum -c
+
+RUN cargo install --git https://github.com/cross-rs/cross
```
Comment (Emilgardis) [inline docker/Dockerfile.cross-in-docker:10]: This is still wrong, it should copy from the checked out code, if this is kept the CI will work, but it'll check the most recent release, not the proposed changes

**Review by Emilgardis — CHANGES_REQUESTED** [review]

File: docker/Dockerfile.cross-in-docker (hunk 4 above)
Comment (otavio) [reply docker/Dockerfile.cross-in-docker:10]: Sure, it should use the checkout for building, thanks @Emilgardis for raising this up.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -170,6 +170,8 @@ jobs:
             - { target: thumbv7em-none-eabi,              os: ubuntu-latest,                    std: 1 }
             - { target: thumbv7em-none-eabihf,            os: ubuntu-latest,                    std: 1 }
             - { target: thumbv7m-none-eabi,               os: ubuntu-latest,                    std: 1 }
+            
+            - { target: cross                             os: ubuntu-latest }
```
Comment (Emilgardis) [inline .github/workflows/ci.yml:174]: This won't work either sadly, it'll have to be it's own task/action for the image to be built and published

**Review by Emilgardis — CHANGES_REQUESTED** [review]
The image needs to be added

**Comment by Alexhuszagh** [comment]
Closed since #878 as merged.

**Pull request closed without merging**